### PR TITLE
feat(merge): allow to always append pr to queue

### PR DESCRIPTION
**Pull request by sileht**

### CODE REVIEW

File: doc/source/actions.rst [hunk 1]
```diff
@@ -351,6 +351,26 @@ The ``merge`` action merges the pull request into its base branch. The
...
+       * ``always-append`` pull requests are always placed at the end of the queue
+
+   * - ``strict_smart_priority_labels``
+     - list of string
+     - `[]`
+     - List of labels to prioritize pull request in queue. Labels must be
+       ordered from the highest priority to the lowest.
+
+       Once a pull request enters in the queue, the priority can be changed even
+       if the label is removed.
```
Comment (jd) [inline doc/source/actions.rst:372]: I'm wondering if we shouldn't do something different, which is by using `conditions`. Labels are good but they are kind limited.

For example you might want PR from a bot to be always low priority. You could do:

```yaml
- name: automatic merge from dependabot
    conditions:
      - author~=^dependabot(|-preview)\[bot\]$
      - "status-success=ci/circleci: py37"
    actions:
      merge:
        strict: "smart"
        priority: -10
        method: rebase
  - name: automatic merge
    conditions:
      - base=master
      - "status-success=ci/circleci: py38"
      - "#approved-reviews-by>=1"
    actions:
      merge:
        strict: "smart"
        method: rebase
        priority: 10
```

I don't like priority being numbered because you never know which is low which is high.
So I'd say make them an `Integer` where the higher the more priority, so it's flexible, plus add some conveninent alias such as:
- `high` = 128
- `normal` = 0
- ` low` = -128

It makes it easy to understand when writing/reading the configuration.

File: doc/source/actions.rst [hunk 2]
```diff
@@ -351,6 +351,26 @@ The ``merge`` action merges the pull request into its base branch. The
...
 
+   * - ``strict_smart_behavior``
+     - string
+     - ``merge-if-ready``
+     - Behavior when a pull request is ready to be merge in ``strict=smart``
+       mode. Possible values:
+
+       * ``merge-if-ready`` if the pull request is ready and already in sync
+         with its base branch, the queue is by-passed and the pull request is
+         directly merged. This behavior allows to safe CIs resources.
```
[the remark this answers is not in the document]
Comment (sileht) [reply doc/source/actions.rst:362]: If you have a long-running CIs, you really save resources.

Assume CI takes 2 hours:

merge-if-ready:
* PR 1 is ready except CIs CI already ran since 1 hour
* PR 2 just become ready and already up2date, we just merge it without queuing it.
* PR 1 get rebased, previous CI RUN get canceled, we wait 2h and  merged it

CIs total runs times 3 hours

always-append:
* PR 1 is ready except CIs, CI already run since 1 hour
* PR 2 just become ready and already up2date, we put it in queue.
* PR 1 CIs finish 1 hour later and get merged
* PR 2 is rebased, CIs rerun and take 2 hours
* PR 2 get merged

CIs total runs times: 4 hours

File: doc/source/actions.rst [hunk 3]
```diff
@@ -351,6 +351,26 @@ The ``merge`` action merges the pull request into its base branch. The
...
+       * ``always-append`` pull requests are always placed at the end of the queue
+
```
[the remark this answers is not in the document]
Comment (sileht) [reply doc/source/actions.rst:364]: good idea !

File: doc/source/actions.rst (hunk 1 above)
Comment (sileht) [reply doc/source/actions.rst:372]: I have thought of this solution too (I have already a branch with that), I prefer it for it's flexibility, but I didn't take this road because I am pretty sure the engine has some race when you have many merge rules. Currently people have no chance to run into them. But if they start switching labels to change the merge rule, this will increase a lot the change to trigger them.

That's said, while I'm writing this, I found a simple idea to fix up the race I have in mind.

I will split this PR to merge the first part and propose the priority thing into another one.

File: doc/source/actions.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (jd) [reply doc/source/actions.rst]: I think you forgot some:

`merge-if-ready`:

- PR 1 is ready except CIs CI already ran since 1 hour (**1 hour CI**)
- PR 2 just become ready and already up2date, we just merge it without queuing it (**2 hours CI**)
- PR 1 get rebased, previous CI RUN get canceled, we wait 2h and merged it (**2 hours CI**)

CIs total runs times **5 hours**

`always-append`:

- PR 1 is ready except CIs, CI already run since 1 hour (**1 hour CI**)
- PR 2 just become ready and already up2date, we put it in queue (**1 hours CI** since it's going to be cancelled)
- PR 1 CIs finish 1 hour later and get merged (**1 hour CI**)
- PR 2 is rebased, CIs rerun and take 2 hours (**2 hours CI**)
- PR 2 get merged

CIs total runs times: **5 hours**

So it's the same, no?

File: doc/source/actions.rst (hunk 1 above)
Comment (jd) [reply doc/source/actions.rst]: Awesome <3

File: doc/source/actions.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (sileht) [reply doc/source/actions.rst]: > I think you forgot some:
> 
> `merge-if-ready`:
> 
>     * PR 1 is ready except CIs CI already ran since 1 hour (**1 hour CI**)
> 
>     * PR 2 just become ready and already up2date, we just merge it without queuing it (**2 hours CI**)

PR 2 here is already rebased, CIs already green, so we do not have this 2 hours

>     * PR 1 get rebased, previous CI RUN get canceled, we wait 2h and merged it (**2 hours CI**)
> 
> 
> CIs total runs times **5 hours**
> 
> `always-append`:
> 
>     * PR 1 is ready except CIs, CI already run since 1 hour (**1 hour CI**)
> 
>     * PR 2 just become ready and already up2date, we put it in queue (**1 hours CI** since it's going to be cancelled)
>     * PR 1 CIs finish 1 hour later and get merged (**1 hour CI**)
> 
>     * PR 2 is rebased, CIs rerun and take 2 hours (**2 hours CI**)
> 
>     * PR 2 get merged
> 
> 
> CIs total runs times: **5 hours**
> 
> So it's the same, no?


Another way to explain it:

In always-append, things are always rebased (except if the queue is empty).
So for a PR, CIs run 2 hours before "merge" action is triggered and then 2 hours after the rebase. For 2 PRs, we have 8 hours

In merge-if-ready, if the PR doesn't need to be rebased, we merge it directly whatever the queue status is and skip the second 2 hours CIs (We still have 2 hours before "merge" action is triggered).
But the counter-part is that the first PR of the queue get its CIs canceled after 1 hour of run, rebased again and then 2 more hours of CIs , so 5 hours in total (2 hours before "merge" action is triggered, 1 hour for nothing, 2 hours after the rebased)
Here, for the two PRs, we have 7 hours in total.

[the remark this answers is not in the document]
Comment (jd) [reply doc/source/actions.rst]: > PR 2 here is already rebased, CIs already green, so we do not have this 2 hours

Of course, you have: for the CI to be green, it has to run for 2 hours. :)

> In always-append, things are always rebased (except if the queue is empty).
So for a PR, CIs run 2 hours before "merge" action is triggered and then 2 hours after the rebase. For 2 PRs, we have 8 hours

👍  that's a good summary. In `always-append`, you always have 2 CI runs; the _election_ CI runs, where you make sure it works and it can enter the queue, and the _merge_ CI run that is done once it's at the top of the queue and going to be merged.

> In merge-if-ready, if the PR doesn't need to be rebased, we merge it directly whatever the queue status is and skip the second 2 hours CIs (We still have 2 hours before "merge" action is triggered).

The thing is that the more you cancel, the more CI time you use for nothing.
In `always-append` the CI time is constant and always _election_+_merge_. In `merge-if-ready`, you always have _election_, but then your _merge_ can be between 0 (the _election_ run is used) or infinite (you keep being canceled).

So it *can* save CI time, but it can also consume CI time forever 😨 😆

File: doc/source/actions.rst [hunk 4]
```diff
@@ -333,6 +333,12 @@ The ``merge`` action merges the pull request into its base branch. The
          against its base branch at a time.
          This allows you to e.g., save CI time, as Mergify will queue the
          mergeable pull requests and update them serially, one at a time.
+         But if the pull request is ready and already in sync with its base
+         branch, the queue is bypassed and the pull request is directly merged
```
Comment (jd) [inline doc/source/actions.rst:337]:
```suggestion
         If the pull request is already in sync with its base branch, the queue is
         bypassed and the pull request is directly merged.
```

File: doc/source/actions.rst [hunk 5]
```diff
@@ -333,6 +333,12 @@ The ``merge`` action merges the pull request into its base branch. The
...
+
+       * ``smart-ordered`` enables :ref:`strict merge`, same behavior as before
+         except queue is never bypassed, pull requests are always added at the
+         end of the queue
```
Comment (jd) [inline doc/source/actions.rst:341]:
```suggestion
       * ``smart-ordered`` enables :ref:`strict merge` with the same behavior as ``smart``,
         except that the queue is never bypassed. Pull requests are always added to the
         end of the queue, even if they are ready to be merged.
```

File: doc/source/actions.rst [hunk 6]
```diff
@@ -333,6 +333,12 @@ The ``merge`` action merges the pull request into its base branch. The
          against its base branch at a time.
          This allows you to e.g., save CI time, as Mergify will queue the
          mergeable pull requests and update them serially, one at a time.
+         But if the pull request is ready and already in sync with its base
+         branch, the queue is bypassed and the pull request is directly merged
+
+       * ``smart-ordered`` enables :ref:`strict merge`, same behavior as before
```
Comment (jd) [inline doc/source/actions.rst:339]:
```suggestion
       * ``smart+ordered`` enables :ref:`strict merge`, same behavior as before
```
I'd prefer we use the `+` symbol instead.

File: mergify_engine/actions/merge/action.py [hunk 7]
```diff
@@ -46,7 +49,9 @@ class MergeAction(actions.Action):
         voluptuous.Required("rebase_fallback", default="merge"): voluptuous.Any(
             "merge", "squash", None
         ),
-        voluptuous.Required("strict", default=False): voluptuous.Any(bool, "smart"),
+        voluptuous.Required("strict", default=False): voluptuous.Any(
+            bool, "smart", "smart-ordered"
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:53]:
```suggestion
            bool, "smart", "smart+ordered"
```

File: mergify_engine/actions/merge/action.py [hunk 8]
```diff
@@ -62,18 +67,30 @@ def run(self, ctxt, rule, missing_conditions):
...
             finally:
-                if self.config["strict"] == "smart":
-                    q.remove_pull(ctxt.pull["number"])
+                q.remove_pull(ctxt.pull["number"])
+        else:
+            return self._sync_with_base_branch(ctxt)
+
+    def _should_be_merged(self, ctxt):
+        q = queue.Queue.from_context(ctxt)
+        if self.config["strict"] in ["smart", "smart-ordered"]:
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:83]:
```suggestion
        if self.config["strict"] in ("smart", "smart+ordered"):
```

File: mergify_engine/actions/merge/action.py [hunk 9]
```diff
@@ -62,18 +67,30 @@ def run(self, ctxt, rule, missing_conditions):
...
+            if self.config["strict"] == "smart-ordered":
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:84]:
```suggestion
            if self.config["strict"] == "smart+ordered":
```

File: mergify_engine/actions/merge/action.py [hunk 10]
```diff
@@ -139,7 +155,7 @@ def _sync_with_base_branch(self, ctxt):
                 "modification",
                 "",
             )
-        elif self.config["strict"] == "smart":
+        elif self.config["strict"] in ["smart", "smart-ordered"]:
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:158]:
```suggestion
        elif self.config["strict"] in ("smart", "smart+ordered"):
```

File: mergify_engine/actions/merge/queue.py [hunk 11]
```diff
@@ -116,8 +116,14 @@ def _move_pull_to_new_base_branch(self, pull_number, old_base_branch):
...
 
+    def is_first_pull(self, pull_number):
+        pull_requests = self.get_pulls()
+        if not pull_requests:
+            return True
+        return pull_requests[0] == pull_number
+
     def get_pulls(self):
-        return self.redis.zrange(self._cache_key, 0, -1)
+        return [int(pull) for pull in self.redis.zrange(self._cache_key, 0, -1)]
```
Comment (jd) [inline mergify_engine/actions/merge/queue.py:126]: why is that required now?

File: mergify_engine/tests/functional/actions/test_merge.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+        self.git("fetch", "--all")
+        p_ready, _ = self.create_pr(base_repo="main")
+
+        self.add_label(p_need_rebase, "ready")
+        self.add_label(p_ready, "ready")
+        return p_need_rebase, p_ready
+
+    def test_merge_smart_ordered(self):
+        p_need_rebase, p_ready = self._do_test_smart_order("smart-ordered")
```
Comment (jd) [inline mergify_engine/tests/functional/actions/test_merge.py:58]:
```suggestion
        p_need_rebase, p_ready = self._do_test_smart_order("smart+ordered")
```

File: mergify_engine/actions/merge/queue.py (hunk 11 above)
Comment (sileht) [reply mergify_engine/actions/merge/queue.py:126]: I just moved where it happens, Set keys has always been a string. Before this was done just after we were calling `get_pulls()`.

Comment (jd) [reply mergify_engine/actions/merge/queue.py:126]: Makes sense!
